### DISCUSSION TITLE
7.13.0@next

### DISCUSSION THREAD

**Discussion by bluebill1049** [react-hook-form/react-hook-form#6117] [General]
We are planning to merge this https://github.com/react-hook-form/react-hook-form/pull/5958 PR by this week. It's a pretty big improvement over performance and internal code structure. We are planning to run the `next` (beta) version for a couple of weeks before release the final version consider this is probably the biggest code change since we released v7. If you are running a complex and large form application, give it a go and give us feedback.

## Steps

We are going to follow the following steps before release the final version.

- [x] merge PR
- [x] Release next version
- [x] Running over could weeks
- [ ] Relase the final version

**Comment by bluebill1049** [comment]
https://github.com/react-hook-form/react-hook-form/releases/tag/v7.13.0-next.0

**Comment by mh-alahdadian** [comment]
I suggest that change useForm returned value and let users to use useFormState to listen on formState it's a better breakdown concept and more important then we do not need to clone formControl object to add a formState field to it [here](https://github.com/react-hook-form/react-hook-form/blob/2d5de49b0fcb8dd6ccea1e70d6f482fcdaff5cce/src/useForm.ts#L104) and that can help us with having a single returned value in different renders which can decrease child re-rendering when passing formControl or when using ContextProvider

**Reply by bluebill1049** [reply]
what do you mean @MHA15?

**Reply by mh-alahdadian** [reply]
```
// function useForm 

return _formControl.current;
```

```
// usage of useForm 
const formControl = useForm();
// use formState hook just if we need state and can not use formControl.formState 
const formState = useFormState({control: formControl.control}); 

return <FormProvider {...formControl}>{tree}</FormProvider>

```
In this way `formControl` will never change between different renders and it cause no Context API update between renders and it can be a huge performance improvement

**Reply by bluebill1049** [reply]
https://github.com/react-hook-form/react-hook-form/pull/6154

**Reply by bluebill1049** [reply]
@MHA15 any feedback on my PR above.

**Reply by mh-alahdadian** [reply]
totally it's good enough , I just add a discussion on it but I think It could be merged before discuss about it
thanks for your work

**Reply by bluebill1049** [reply]
thanks @MHA15

**Comment by TonyVeigel** [comment]
On 7.13.0 next, when using field array, if you use reset and then try to set value with shouldDirty flag equal too true, isDirty is still false. 

Example here: https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-forked-2dtey?file=/src/index.js 

Click reset, then click should dirty button -> doesn't work, isDirty is still false.
Don't click reset and just click should dirty -> works, isDirty is true.

**Reply by bluebill1049** [reply]
Does it work with preivous version?

**Reply by TonyVeigel** [reply]
Yup, we have a large scale app using field array and we upgraded to the next version to see how it works and noticed that as soon as we upgraded that functionality broke.

**Reply by bluebill1049** [reply]
looking into this issue.

**Reply by bluebill1049** [reply]
thanks fixed in new `next` release: https://github.com/react-hook-form/react-hook-form/releases/tag/v7.13.0-next.1

**Comment by bluebill1049** [comment]
https://github.com/react-hook-form/react-hook-form/releases/tag/v7.13.0-next.1 published

**Comment by kamqo** [comment]
This is not strictly related to new v7.13.0 release but I think it should be fixed soon as it might influence multiple people.
Referencing issue - behaviour of `shouldUnregister` that was implemented in version v7.6.0 and above as far as I understood it correctly. Firstly I wanted to report it as bug but it would need some time to prepare demo for demonstration and perhaps it won't be necessary.

Issue:
We have our form with `shouldUnregister: true`.
Let's pretend we have dynamic DOM element structure where we register fields based on other react elements logic.
We have 3 components A,B,C where A has logic for displaying B and C.
If A property is equal to `abc` then show component B,
if A property is equal to `fgh` then show component C. 

What if component B and C share same input field (same register name). e.g. `<input type="text" ...register("test")>` with same name `test`. Then when you switch condition in component A property from `abc`->`fgh` it will not result in unregistering `test` and therefore it lets previous value from B present in component even when we are displaying component C.

-------------------------

Firstly I thought that react has some kind of memo mechanism but actually it calls component unmount method therefore there isn't issue. Next test I did was to check if custom unregister via useLayoutEffect works without relying on `shouldUnregister` setting and it works, therefore I assume you implemented different mechanism when you reset internal state of form when form field is removed.
```
    useLayoutEffect(() => () => {
        // unmounted element.
        console.log("unmounted - " + props.name);
        // I would expect component will call unregister when is unmounted from form.
        unregister(props.name);
    }, [])
```
If you need specific demo or any other information regarding that, let me know but somehow it blocks our development as we works with multiple templates and have highly dynamic structure.

**Comment by bluebill1049** [comment]
plan to release 7.13.0 next week.

**Comment by bluebill1049** [comment]
released.
